Distribution power flow and test cases.

todorovski-m opened the pull request.

rdzman commented:
:thumbsup: Nice work, Mirko. From a quick first pass, here are a few comments, all open to discussion, and in no particular order:

1. I don't think we need to add `case33bw_radial.m`. It appears to be identical to `case33bw.m` with the exception that the rows of `branch` corresponding to the tie lines are commented out rather than simply having their `BR_STATUS` column set to 0.

2. Please use the named constants defined by `idx_bus`, `idx_gen`, etc. rather than the numerical values when indexing the columns of the data matrices or for things like `BUS_TYPE` values. This make the code more robust to potential future updates to the case file format.

3. It would be nice to have some more verbose progress output, e.g. printing mismatch magnitudes per iteration (see Newton method, for example), when the `verbose` option is greater than 1. You could even be more verbose for `verbose = 3`.

4. I would like to see a complete set of tests comparing the outputs (full `bus`, `gen`, and `branch` matrices) of each of the radial power flow algorithms to the results of the Newton method. See `t_pf.m` for a model. In this case, rather than comparing against a saved solution, simply run a Newton power flow as well and compare against that solution. If you did it for each of the included radial networks as well, it would also serve as a minimal test of those data files too. Call the test function `t_pf_radial` and add it to `test_matpower`.

5. Make sure you use Unix line endings for your files. I noticed in several files some red `^M` characters at the end of some lines when scanning through the diffs shown [here](https://github.com/MATPOWER/matpower/pull/5/commits/17dc4129b0dbe9dd59563b0d1d1124b7384a7660). If you're working on a Windows machine, there is apparently a Git setting to _"Checkout Windows-style, commit Unix-style line endings"_.

6. It seems that `order_radial` might best be implemented as an option to `ext2int`, since it really is simply creating an specialized internal ordering of the buses and branches. Also, it is essentially doing a traversal of the network, I believe similar to what is done in `find_islands`. I'd rather avoid duplication of functionality and I'm concerned that the approach in `order_radial` might be significantly slower on large systems. Maybe there is a way to refactor things so both functions use the same underlying code. On a related note, I've wanted to modify `find_islands` to return a `radial` flag for each connected component of the graph. Maybe we could modify it to do that and also to return the ordering information needed by an `order_radial` option to `ext2int`. We can discuss whether it makes sense to do this now, or work on it as a separate enhancement after merging this pull request.

todorovski-m commented:
Ray, here are the answers on the comments:

1. I have deleted `case33bw_radial.m`. I did not notice that the branches forming loops were with  `BR_STATUS` column set to 0.
2. All code now use the named constants.
3. Verbose progress output was added printing mismatch in voltage magnitudes per iteration.
4. I have created `t_pf_radial.m` where I check the results obtained with Newton method and all three new methods. The test are performed for each radial test case. In all tests mismatches are less then `1e-6`. I have added `t_pf_radial` to `test_matpower`. The test can also be run via `all_ok = t_run_tests({'t_pf_radial'},1)`
5. Some of the files where earlier created on a Windows machine (now I'm using Linux). I have converted them using the utility `dos2unix` and now red `^M` characters at the end of lines are not present.
6. I agree that we should avoid  duplication of functionality. The ordering I'm using has to ensure the following: slack bus must have index 1, branches should have “to” indexes bigger then “from” indexes. I think that this may come out of `find_islands` but I have to study it more deeply. This is not address yet.

todorovski-m commented:
Distribution power flow is updated to handle branches with taps different then 0 (transformers). In `case4_dist`, which is a hypothetical one I have set `TAP` column of the last branch to 1.05. Other distribution cases, which are taken from the literature, are not changed.

rdzman commented:
I've made a few changes here in preparation for merging.
- merged in the latest changes from upstream (now includes `install_matpower` to set paths)
- moved all case files to a `data` directory
- updated the `mpoption` version to 13 and included code to convert v12 to v13.

There are a few other things I was wondering if you could handle before we merge this.
- Currently, if you select one of these methods with a meshed system, it gives you an `Index exceeds matrix dimensions` error. Could you have it do an explicit check before that happens and throw an explicit error that says something like ...

      radial_pf: power flow algorithm 'PQSUM' can only handle radial networks

- In the tests, could you add testing of `pf.radial.vcorr = 1` and `pf.radial.vcorr = 0` and also test the number of iterations against a specified array of expected values, which I expect may be the main difference for the two values of that option.
  - You'll need to use a `results` struct rather than individual matrices for the return value of `runpf`.
  - It may be simplest to set the options and test labels dynamically (create a cell arrays of labels and algorithms and loop through them).
- Finally, I'll need to figure out how I'm going to handle the User's Manual, before things get too out-of-sync. Would you mind writing a "Radial Power Flow" section that we could put in as section 4.1.1 maybe? Be sure to include relevant references. It doesn't have to be long. I'm thinking of something between a couple of paragraphs and a page or two. I can take care of integrating it if I don't get around to making the manual source public before we're ready to merge this.

Thanks again.

rdzman commented:
I just merged in the latest changes from upstream. The LaTeX source for the manual is now included in `docs/src/MATPOWER-manual/MATPOWER-manual.tex`.

todorovski-m commented:
- I have added check in `radial_pf` which simply compares number of buses and branches. For radial networks the difference should be 1. If this is not the case `radial_pf` prints an error message and quits.

- `t_pf_radial` is modified so that it performs two sets of tests. In the first set the tests are as before, since there are no PV buses in the test cases there is no need to test with  `pf.radial.vcorr = 1` and `pf.radial.vcorr = 0`. Voltage correction procedure is performed at PV buses only. In the second set of test I have added PV buses to all test cases. I added 5 PV buses at locations with 5 lowest voltages in the original case. The generator voltage at these buses is set 0.05 pu higher then the voltage calculated with NR in the original case. At the end `t_pf_radial` prints the number of iterations performed by each method for each test case.

- Finally, of course I will be glad to write a section on "Radial Power Flow". I believe I can do it in a week.

rdzman commented:
Thanks. I have a few comments on the code changes, but first I'm trying to figure out what happened with my and your latest commits. I'm relatively new to git, so I might be misunderstanding what happened, but did you `--force` push some changes? It seems like the history of your `master` branch (on which the pull request is based) got rewritten and no longer includes some of my commits listed in this thread (e.g. 4b224ab, f789308, cf2c899). Seems like several of my commits got rewritten and squashed into your last merge commit d56b652. I don't think that should happen unless you are running into merge conflicts and then `--force` pushing things.

I'm sure we can sort it out, but I just want to make sure I understand what's going on so we can be sure we each use a workflow that doesn't cause problems for the other.

todorovski-m commented:
No, I did not `--force` push changes. I just did `git push origin master` like before. Then I got a message that I have to perform `git pull` first since there were changes which were not incorporated in my directory. I'm new to `git` too. In the latest push I pushed only 2 files: `radial_pf.m` and `t_pf_radial.m`. From `git log` I got the following for the last push:

commit d56b6527fa34ad83516646a6e4ba52ddbc162ba4
Merge: 2a2505a 9892582
Author: Mirko Todorovski <todorovski.mrk@gmail.com>
Date:   Fri Jan 27 21:45:24 2017 +0100

    Merge branch 'master' of https://github.com/todorovski-m/matpower

commit 2a2505a2ea6b1ab7048a623e3b472827c8fa9638
Author: Mirko Todorovski <todorovski.mrk@gmail.com>
Date:   Fri Jan 27 21:39:32 2017 +0100

    Improved radial power flow testing.

rdzman commented on lib/radial_pf.m:
```diff
@@ -0,0 +1,94 @@
+%  See also CASEFORMAT, LOADCASE, MPOPTION.
+
+%% load case data
+mpc = loadcase(casedata);
```
Since `radial_pf()` is called only from `runpf()` and requires a MATPOWER case struct that has already been converted to internal bus numbering, this line is not needed. In fact, I suggest changing the name of the first argument from `casedata` to `mpc` and document it as a MATPOWER case struct with internal bus numbering.

todorovski-m replied:
I have changed first argument from `casedata` to `mpc` and stated that it as case struct with internal bus numbering.

rdzman replied:
👍

rdzman commented on lib/radial_pf.m:
```diff
@@ -0,0 +1,94 @@
+mpc = loadcase(casedata);
+nb = size(mpc.bus,1);
+nl = size(mpc.branch,1);
+if nb - nl ~= 1
```
This is a necessary, but not sufficient condition for a radial network. I don't remember where the error happened, but I thought it was during the ordering. Can't you check for some condition there to more explicitly check for a radial network?

todorovski-m replied:
The test whether a network is radial is performed in `order_radial` which now returns a field `mpc.loop` with indices of branched that are forming loops. If `verbose = 1` function `radial_pf` prints a list of branches in `mpc.loop`.

rdzman replied:
Good. I forgot to mention that I think it should be a fatal error (i.e. use `error()` instead of `fprintf()` and `success = 0`) when the network is not radial. Also, as a best practice, I try never to use `return` in the middle of a function.

todorovski-m replied:
Done.

rdzman commented on lib/t/t_pf_radial.m:
```diff
@@ -0,0 +1,138 @@
+end
+
+t_end;
+
```
The printing below should only happen if `verbose` is true. Also, my idea was to take the results of this and hard code it in a cell array, then include a `t_is` that checks `r.iterations` against the corresponding hard-coded value for each run. That way if someone changes the code in a way that causes the algorithm to behave differently, the iteration counts may change, and tests may fail. The goal is to have inadvertent changes show up as failed tests.

todorovski-m replied:
Expected number of iterations are hard coded in `t_pf_radial` and for each test `r.iterations` is checked against hard coded values.

rdzman replied:
👍 

Except I'd just use `if verbose` or `if verbose > 0` rather than `if verbose == 1`. It's just for consistency, since throughout the MATPOWER code, we assume that `verbose` can take on values greater than 1 as well.

todorovski-m replied:
Done.

rdzman commented:
Hmmm ... ok. No worries.

Well, I'd like to be able to keep a nice linear commit history on pull request branch if possible. Maybe the best way to do that for future modifications is (for each of us) to do any local modifications on a separate branch, then before pushing anything to the shared pull-request branch, make sure the pull-request branch is up-to-date, then rebase our new changes on top of the pull-request branch before merging them so they enter as a simple fast-forward merge.

rdzman commented:
Go ahead and merge your latest changes into your `master` branch so they are included in this pull request. That should be a fast-forward merge since I haven't made any changes. I think the only thing remaining then is the additions to the User's Manual.

todorovski-m commented:
I have created new branch `radial_pf.tex` which consist of an initial version of subsection on radial power flow. It is not polished and not included in the big Matpower manual.

todorovski-m commented:
Out of curiosity, I have created a new branch `find_islands` with a function `find_islands_1` to detect islands and return cell array `GROUPS` with vectors of the bus indices for each island and vector `ISOLATED` with bus indices of isolated buses. There is a paper `islands_detection.pdf`, which was once submitted for publication and rejected, explaining the procedure.

todorovski-m commented:
I have added branch `loops-bug` with changes in `order_radial`. I've spotted a problem when trying it with `case6ww`. When building the spanning tree after first step it consists of branches 1-2, 1-4, 1-5 and 3-5. Then in the second step candidates for the tree are 2-6, 3-6 and 5-6. If I add them all they will form 2 undetected loops (3-6 and 5-6). I have now changed the code so that it adds only 1 branch going to a newly inserted bus, while the others are declared loops.
I believe it is safe to merge branch `loops-bug` with `master`.

todorovski-m commented:
In the new commit to `loops-bug` I added slightly modified (improved) method for handling PV nodes in backward/forward power flow. It is based on the following [paper.](http://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=964896&isnumber=20826) The number of iterations is lowered.

rdzman commented:
Thanks again for all your work on this Mirko.

I merged the changes from your `loops-bug` branch into this pull request (your `master` branch) and incorporated the section for the manual into the main manual itself, along with numerous edits. I think this pull request is now ready for merging into MATPOWER's `master` branch. Let me know if you have anything else you'd like to include before I do that.

If you want to pursue the discussion regarding your `find_islands` code, we can take that up as a separate issue or pull request. It wasn't clear to me if you were proposing to replace the current `find_islands()` function with  your version. I did test it on a large system and it was quite slow (16 sec for a 62,000 bus model, vs. 0.4 sec for the existing `find_islands()`.

todorovski-m commented:
For now I do not have anything else to include. I'm planning to modify the code to solve cases with loops, but that is another story.
Concerning `find_islands_1` I suggest to discuss it separately, later. It is not tested and compared with original `find_islands`.

rdzman commented:
👏  Nice work. It's been merged. I did a squash and merge, so all of these changes entered as a single commit.

I believe it includes everything from your `loops-bug`, `ordering+t_pf_radial`, and `radial_pf.tex` branches, so feel free to delete them. Also, I suggest that you now reset [`todorovski-m/matpower:master`](https://github.com/todorovski-m/matpower/tree/master) to be in sync with [`MATPOWER/matpower:master`](https://github.com/MATPOWER/matpower/tree/master), and in the future keep it in sync, always creating a new branch off of that for any future pull requests.

Thanks again for this valuable contribution! 👍 😃 ⭐️   Very much appreciated! 👏

todorovski-m commented:
Thank you, it was pleasure working with you.
I deleted `loops-bug`, `ordering+t_pf_radial`, and `radial_pf.tex` branches locally. I also like to reset [todorovski-m/matpower:master](https://github.com/todorovski-m/matpower/tree/master) to be in sync with [MATPOWER/matpower:master](https://github.com/MATPOWER/matpower/tree/master) but I'm not sure how to do that. I'm afraid that I will spoil something. Could you please help me.

rdzman commented:
You can feel free to delete those branches on GitHub as well, either via the web interface, or via the command-line (assuming `origin` points to your GitHub fork, which I assume it does):

    git push origin --delete loops-bug
    git push origin --delete ordering+t_pf_radial
    git push origin --delete radial_pf.tex

Btw, did you set up Git as described in [CONTRIBUTING.md](/MATPOWER/matpower/blob/master/CONTRIBUTING.md)? In particular, do you have a remote called `upstream` that points `https://github.com/MATPOWER/matpower.git`? You can check with:

    git remote -v

If so, you should see something like …

    origin      https://github.com/todorovski-m/matpower.git (fetch)
    origin      https://github.com/todorovski-m/matpower.git (push)
    upstream    https://github.com/MATPOWER/matpower.git (fetch)
    upstream    https://github.com/MATPOWER/matpower.git (push)

If not, you can add it using …

    git remote add upstream https://github.com/MATPOWER/matpower.git

Then, once you have `origin` and `upstream` pointing to the right remotes, the following should reset your `master` branch to be in sync with the `upstream` `master` branch.

    git checkout master
    git fetch upstream
    git reset --hard upstream/master  
    git push origin master --force 

At that point, the `master` branch of your fork should be back in sync with the `upstream` repo and you can rebase anything else you may be working on on top of it.

    git checkout <topic-branch-name>
    git rebase master
    git push origin <topic-branch-name>         (may require --force)
    (create pull request of <topic-branch-name>, if/when ready)

Let me know if you have any trouble.

todorovski-m commented:
Everything is OK. On the GitHub web interface for my `master` now I see "This branch is even with MATPOWER:master." Thank you for your help.

akhilazm commented:
Is there any way to get the maximum load limit of a distribution system as a whole. Running radial power flow solution for increasing load at each PQ buses do not converge even changing the voltage limit (from 1.1 and 0.9) provided in mpc.bus struct case.
Please help me to get out of this.
